### PR TITLE
Variable number of image cells, lazy default hash

### PR DESCRIPTION
**Pull request by programWhiz**
This fixes two issues:

1. Instead of hardcoding 5 cells, this just allows a variable number of cells using `numCells` option.
2. Performance: instead of always computing a default hash, only compute it if the hash isn't given.

### CODE REVIEW

**Comment by stewartlord** [comment]
Hi, just taking a look at this PR and I'm a bit unclear on what effect num-cells is supposed to achieve. It will produce a different looking image, but without other changes it is still the same number of cells that will be drawn, just in different positions.

**Comment by programWhiz** [comment]
Right you are, looks like this part needs a change as well:
```
var i, color;
            for (i = 0; i < 15; i++) {
                color = parseInt(this.hash.charAt(i), 16) % 2 ? bg : fg;
                if (i < numCells) {
                    this.rectangle(2 * cell + margin, i * cell + margin, cell, cell, color, image);
                } else if (i < 10) {
                    this.rectangle(1 * cell + margin, (i - numCells) * cell + margin, cell, cell, color, image);
                    this.rectangle(3 * cell + margin, (i - numCells) * cell + margin, cell, cell, color, image);
                } else if (i < 15) {
                    this.rectangle(0 * cell + margin, (i - 10) * cell + margin, cell, cell, color, image);
                    this.rectangle(4 * cell + margin, (i - 10) * cell + margin, cell, cell, color, image);
                }
            }
```

**Comment by programWhiz** [comment]
Ok that should fix it.  Also now the code is simpler.  Algorithm works like this now:
1. Iterate each row, and half the columns
2. Choose color for <row, col> (approximately what you were doing before)
3. Draw cell for the <row, col>, mirror about the center
screen shot 2017-01-24 at 6 51 16 am
screen shot 2017-01-24 at 6 51 22 am

File: identicon.js [hunk 1]
```diff
@@ -33,11 +33,19 @@
         if (typeof(arguments[1]) === 'number') { this.options.size   = arguments[1]; }
         if (arguments[2])                      { this.options.margin = arguments[2]; }
 
-        this.hash        = hash                    || this.defaults.hash;
+        this.hash        = hash                    || this.defaultHash();
         this.background  = this.options.background || this.defaults.background;
         this.margin      = this.options.margin     || this.defaults.margin;
         this.size        = this.options.size       || this.defaults.size;
         this.format      = this.options.format     || this.defaults.format;
+        this.numCells    = this.options.numCells   || this.defaults.numCells;
```
Comment (stewartlord) [inline identicon.js:41]: I don't think `numCells` is an accurate name for what this variable represents. Is it perhaps `squareSize`?

File: identicon.js [hunk 2]
```diff
@@ -62,26 +71,23 @@
...
-                    this.rectangle(4 * cell + margin, (i - 10) * cell + margin, cell, cell, color, image);
-                }
-            }
+            var halfWidth = (numCells-1) / 2;
+            var i, j, hashIdx;
+            for( i = 0; i < numCells; ++i ){
+              for( j = 0; j <= halfWidth; ++j ){
+                // Different possible color for each column in image
+                hashIdx = (i * halfWidth + j) % this.hash.length;
+                color = parseInt(this.hash.charAt(hashIdx), 16) % 2 ? bg : fg;
```
Comment (stewartlord) [inline identicon.js:86]: What happens if `hashIdx` exceeds the length of the hash?

File: identicon.js [hunk 3]
```diff
@@ -62,26 +71,23 @@
...
-                    this.rectangle(2 * cell + margin, i * cell + margin, cell, cell, color, image);
-                } else if (i < 10) {
-                    this.rectangle(1 * cell + margin, (i - 5) * cell + margin, cell, cell, color, image);
-                    this.rectangle(3 * cell + margin, (i - 5) * cell + margin, cell, cell, color, image);
-                } else if (i < 15) {
-                    this.rectangle(0 * cell + margin, (i - 10) * cell + margin, cell, cell, color, image);
-                    this.rectangle(4 * cell + margin, (i - 10) * cell + margin, cell, cell, color, image);
-                }
-            }
+            var halfWidth = (numCells-1) / 2;
```
Comment (stewartlord) [inline identicon.js:80]: Could you add a comment briefly explaining how this drawing algorithm works?

File: identicon.js (hunk 2 above)
Comment (stewartlord) [inline identicon.js:86]: It looks like `color` is no longer being declared.

File: identicon.js [hunk 4]
```diff
@@ -62,26 +71,23 @@
...
+            var i, j, hashIdx;
```
Comment (stewartlord) [inline identicon.js:81]: Instead of `hashIdx`, perhaps `hashIndex`

**Review by stewartlord — CHANGES_REQUESTED** [review]
Thanks! Back to you :)

**Pull request closed without merging**